sbat.c

```diff
@@ -120,4 +121,168 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
+                       entry->component_name);
+                /* atoi returns zero for failed conversion, so essentially
+                badly parsed component_generation will be treated as zero
+                */
+                sbat_gen = atoi(entry->component_generation);
+                sbat_var_gen = atoi(sbat_var_entry->component_generation);
+                if ( sbat_gen < sbat_var_gen ) {
+                        dprint(L"component %a, generation %d, was revoked by SBAT variable",
+                                 entry->component_name, sbat_gen);
+                        LogError(L"image did not pass SBAT verification\n");
+                        return EFI_SECURITY_VIOLATION;
+                }
+        }
+        return EFI_SUCCESS;
+}
+
+EFI_STATUS
+verify_sbat(struct sbat *sbat, list_t *sbat_entries)
+{
+	unsigned int i;
+	list_t *pos = NULL;
+	EFI_STATUS efi_status;
+	
+	for (i = 0; i < sbat->size; i++) {
+		struct sbat_entry *entry = entry = sbat->entries[i];
+		for_each_sbat_var(pos, sbat_entries) {
+			struct sbat_var *sbat_var_entry = list_entry(pos, struct sbat_var, list);
+			efi_status = verify_single_entry(entry, sbat_var_entry);
+			if (EFI_ERROR(efi_status))
+				return efi_status;
+		}
+	}
+	dprint(L"all entries from SBAT section verified\n");
+	
+	return efi_status;
+}
+
+static BOOLEAN
+is_utf8_bom(CHAR8 *buf, size_t bufsize)
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+	return !CompareMem(buf, bom, MIN(sizeof(bom), bufsize));
+}
+
+static struct sbat_var *
+new_entry(const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+
+	INIT_LIST_HEAD(&new_entry->list); 
+	new_entry->component_generation = comp_gen;
+	new_entry->component_name_size = comp_name_size;
+	new_entry->component_name = comp_name;
+	return new_entry;
+}
+
+EFI_STATUS
+add_entry(list_t *list, const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new;
+	new = new_entry(comp_gen, comp_name_size, comp_name);
+	if (!new)
+		return EFI_OUT_OF_RESOURCES;
+
+	list_add_tail(&new->list, list);
+	return EFI_SUCCESS;
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	for_each_sbat_var_safe(pos, tmp, entries) {
+		struct sbat_var *entry = list_entry(pos, struct sbat_var, list);
+		list_del(&entry->list);
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_generation);
+		if (entry->component_name_size)
+			FreePool((CHAR8 *)entry->component_name_size);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_name);
+		FreePool(entry);
+	}
+}
+
+EFI_STATUS
+parse_sbat_var(list_t *entries)
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+
+	INIT_LIST_HEAD(entries); 
+	efi_status = get_variable(L"SBAT", &data, &datasize, SHIM_LOCK_GUID);
+	if (EFI_ERROR(efi_status)) {
```

Comment:
`if (EFI_ERROR(efi_status) && efi_status != EFI_NOT_FOUND) {`

Reply:
for `EFI_NOT_FOUND` you should just print a message and return `EFI_SUCCESS` I believe, since as you mentioned there may not be a variable at the beginning.